ssl/t1_lib.c

```diff
@@ -2639,30 +2660,130 @@ static int has_usable_cert(SSL *s, const SIGALG_LOOKUP *sig, int idx)
                 continue;
 
             ERR_set_mark();
-            supported = EVP_PKEY_supports_digest_nid(s->cert->pkeys[idx].privatekey,
-                                                     mdnid);
+            supported = EVP_PKEY_supports_digest_nid(pkey, mdnid);
```

Comment:
I'm rather confused about this chunk (and thus commit 2d263a4a73f852005b16359873475d48755999ad).  The intent of this function is to look at the certificate and corresponding private key that we're thinking about using as our EE cert/signature key, and decide whether we think it can produce a signature on CertificateVerify that the client will accept.  The specific call to X509_get_signature_info() is to look at the signature that our CA made on the certificate we are considering using, and to see what algorithm was used to make that signature.  If the client said that it can't validate certificate signatures using that algorithm, then we'll skip that certificate.  But this check looks to be considering a property of our private key, the one we would use to sign CertificateVerify, and comparing that against the digest that was used to sign our own certificate (by the CA).  Those are completely different signatures!
I think perhaps David was running into the bug that we're fixing here, and applied a fix in the wrong place.
But I don't think we need to make any comparison between our (potential) private key and the message digest used to sign our certificate.

On the other hand, the code you've added above in tls1_check_sig_alg() may now fully supersede this (minimal) usage of the peer_cert_sigalgs, since that code will check all the (non-leaf) certs in the chain as having a public key that makes the acceptable type of signature, and this was just looking at the first (last?) signature in the chain.

Reply:
Hmmm. Yes - I think you're right.

Ping @dwmw2

Reply:
> . But this check looks to be considering a property of our private key, the one we would use to sign CertificateVerify, and comparing that against the digest that was used to sign our own certificate (by the CA). Those are completely different signatures!

That appears to be true. 

On reflection, I cannot recall why I went from the initial code attempt in https://github.com/openssl/openssl/issues/7348#issuecomment-430058543 to a different implementation in `has_usable_cert()` by the time I posted the PR in #7408.

The issue was as described in #7348 — I have a hardware-based EVP_PKEY which cannot support certain hash algorithms, and needed to refrain from attempting to do so.

I can't actually see why, and am on vacation so can't spend much time investigating right now, but the patch in question did actually work.

Reply:
I'm having difficulty reproducing this right now; I can't get [TPMv2 ENGINE](https://git.kernel.org/pub/scm/linux/kernel/git/jejb/openssl_tpm2_engine.git/) to work at all with RSA keys, without complaining 'Non PKCS1 padding asked for'. But building different versions of OpenSSL and building and installing engines against them mostly exceeds my limited attention span this week.

The original issue should be reproducible with any EVP_PKEY which refuses to sign a SHA512 signature, and which says so in its `EVP_PKEY_supports_digest_nid()` function. We could do that in software without having to use TPMv2 or other hardware. Shame on me for not adding a test case when I did it...